File: 19.remove_nth_node_from_end_of_list/removeNthFromEnd.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
function ListNode(val){
    this.val = val;
    this.next = null;
}
var removeNthFromEnd = function(head, n) {
    var dummyNode = new ListNode(-1);
    dummyNode.next = head;
    var dummyNode2 = dummyNode;
    var length = 1;
    //get the length of the list code
    while(dummyNode2.next !== null){
        dummyNode2 = dummyNode2.next;
        length++;
    }
    var dummyNode3 = dummyNode;
    for(var i=0; i < length - n - 1 ; i++){
        dummyNode3 = dummyNode3.next;
    }
    var deleteNode = dummyNode3.next;
    dummyNode3.next = deleteNode.next;
    return dummyNode.next;
};

File: 19.remove_nth_node_from_end_of_list/removeNthFromEnd.py
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummy = ListNode(-1)
        dummy.next = head
        slow = fast = dummy
        while n:
            if fast:
                fast = fast.next
                n -= 1
            else:  # invalid case
                return None
        while fast and fast.next:
            fast = fast.next
            slow = slow.next
        slow.next = slow.next.next
        return dummy.next
